plugins/api-docs/src/components/ApiDefinitionDialog/ApiDefinitionDialog.tsx

```diff
@@ -0,0 +1,189 @@
+    height: '100%',
+    backgroundColor: theme.palette.background.paper,
+  },
+  tabs: {
+    borderRight: `1px solid ${theme.palette.divider}`,
+    flexShrink: 0,
+  },
+  tabContents: {
+    flexGrow: 1,
+    overflowX: 'auto',
+  },
+  title: {
+    color: theme.palette.text.primary,
+    wordBreak: 'break-word',
+    fontSize: theme.typography.h3.fontSize,
+    marginBottom: 0,
+  },
+  type: {
+    textTransform: 'uppercase',
+    fontSize: 11,
+    opacity: 0.8,
+    marginBottom: theme.spacing(1),
+    color: theme.palette.text.primary,
+  },
+}));
+
+function TabPanel(props: {
+  children?: React.ReactNode;
+  index: number;
+  value: number;
+}) {
+  const { children, value, index, ...other } = props;
+  const classes = useStyles();
+  return (
+    <div
+      role="tabpanel"
+      hidden={value !== index}
+      id={`vertical-tabpanel-${index}`}
+      aria-labelledby={`vertical-tab-${index}`}
+      className={classes.tabContents}
+      {...other}
+    >
+      {value === index && (
+        <Box pl={3} pr={3}>
+          {children}
+        </Box>
+      )}
+    </div>
+  );
+}
+
+function a11yProps(index: number) {
+  return {
+    id: `vertical-tab-${index}`,
+    'aria-controls': `vertical-tabpanel-${index}`,
+  };
+}
+
+/**
+ * A dialog that lets users inspect the API definition.
+ *
+ * @public
+ */
+export function ApiDefinitionDialog({
+  open,
+  entity,
+  onClose,
+}: {
+  open: boolean;
+  entity: ApiEntity;
+  onClose: () => void;
+}) {
+  const classes = useStyles();
+  const [activeTab, setActiveTab] = React.useState(0);
+
+  useEffect(() => {
+    setActiveTab(0);
+  }, [open]);
+
+  const config = useApi(apiDocsConfigRef);
+  const { getApiDefinitionWidget } = config;
+
+  if (!entity) {
+    return <Alert severity="error">Could not fetch the API</Alert>;
+  }
+
+  const definitionWidget = getApiDefinitionWidget(entity);
+  const entityTitle = entity.metadata.title ?? entity.metadata.name;
+
+  let index = 0;
+
+  return (
+    <Dialog
+      fullWidth
+      maxWidth="xl"
+      open={open}
+      onClose={onClose}
+      aria-labelledby="api-definition-dialog-title"
+      PaperProps={{ className: classes.fullHeightDialog }}
+    >
+      <DialogTitle id="api-definition-dialog-title" disableTypography>
+        <Typography className={classes.type}>
+          API - {definitionWidget?.title ?? 'Raw'}
+        </Typography>
+        <Typography className={classes.title} variant="h1">
+          {entityTitle}
+        </Typography>
+      </DialogTitle>
+      <DialogContent dividers>
+        <div className={classes.root}>
+          <Tabs
+            orientation="vertical"
+            variant="scrollable"
+            value={activeTab}
+            onChange={(_, newValue) => setActiveTab(newValue)}
+            aria-label="Inspector options"
+            className={classes.tabs}
+          >
+            {definitionWidget ? (
+              <Tab label={definitionWidget.title} {...a11yProps(index++)} />
```

Comment:
Inconsistent use of `++` below - probably better to _always_ use `++` since on line 174 i actually think it risks getting the wrong value sometimes. Either way, makes it easier to reason about.

Could also make the a11yProps a wrapped thing that deals with the index internally :)

```ts
function makeA11yProps() {
  let index = 0;
  return () => ({
    index: index++,
    ...
  })
}
```

then make one of those at the top of the function. It'll make sure to auto increment internally every time you call it. And you can do stuff like `index={a11yProps().index}` if that's all you wanted

Reply:
Tab indices and A11y got a little bit messed up here. I refactored it to be more explicit.